elastic/fleet-server
# Using enrollment Id in agent enroll to delete duplicate agent on failed enroll

## What is the problem this PR solves?

In scale tests we were seeing duplicate agents when the enrollment failed for the first time. This pr fixes that. 

## How does this PR solve the problem?

Support `enrollment_id` in enrollment request, which is a unique enrollment identifier.  When fleet server finds the `enrollment_id` in `.fleet-agents`, it deletes the old agents and invalidates the old API key.

This solves the duplicate agent issue in case of a failed first enrollment (for example horde didn't receive the enrollment response for the first time and tries to enroll again).


## How to test this PR locally

- call enroll API twice with the same `enrollment_id`
- verify that one agent is enrolled at the end with the latter `enrolled_at` time and healthy

```
   curl -sk -XPOST -H 'content-type:application/json'  \
   -H "Authorization: ApiKey myapikey" \
   http://localhost:8221/api/fleet/agents/enroll  -d '
   {
      "type": "PERMANENT",
      "enrollment_id": "1234",
      "metadata": {
         "user_provided": {},
         "local": {},
         "tags": []
      }
   }' 
```

TODO integration test, changelog

## Design Checklist




## Checklist





## Related issues

Closes #2254
